attribute-projections/src/main/java/org/hypertrace/core/attribute/service/projection/AttributeProjection.java

```diff
@@ -0,0 +1,12 @@
+package org.hypertrace.core.attribute.service.projection;
+
+import java.util.List;
+import org.hypertrace.core.attribute.service.v1.AttributeKind;
+import org.hypertrace.core.attribute.service.v1.LiteralValue;
+
+public interface AttributeProjection {
+
+  AttributeKind getArgumentKindAtIndex(int index);
```

Comment:
Does this need to be exposed? Is it rather better to have a wrapper class for arguments so that the implementations can have metadata around the actual attribute also?

Reply:
I could potentially be convinced here, but reframing the question, it's still a matter of: Should the projection executor code (currently trace client) be responsible for coercing data to the right format, or should each projection get data in and be responsible for doing so internally. I tend to prefer the former because it feels like that coercing is more about data types than any specific implementation so would often be duplicated, but would gladly discuss.

Reply:
So, looking at operators (PROJECTION_OPERATOR_DEFAULT, PROJECTION_OPERATOR_CONCAT, etc), I thought the result type is implicit there, and hence the type might be internal to Projection was my understanding. It's true in first two cases where a Project is either an existing attribute or Literal because it knows the type. If ProjectionExpression also knows the type, everything will be uniform. wdyt?

Reply:
Not sure I follow - some operators will require strings to work with, but could take in a numeric attribute - for example if I wanted to hash a numeric and string together - will always produce a string value. There, the numeric input needs to be made into a string. Other operators, like DEFAULT don't care at all what the argument type is, so in those cases no conversion needs to be done/attempted. The return type depends on what comes in. In other words, you can default either a string or numeric field.

Reply:
My point is, project should know the incoming types because it has the arguments and their attribute metadata and it knows the output time so the conversion can be internal to it. Let me know if I'm missing something.

Reply:
It certainly can be implemented internally, I just am missing what is gained by doing so. We have information on either side, so we'll need a back + forth communication either way. Currently, the executor 
1. Gets all the input values according to the definition - these can be a mixture of attributes that are pulled out of the span, literal values, and results of other projections.
2. Asks the operation what type it expects for each of those input values and converts as appropriate, erroring out if any can not be converted to the appropriate type. (this would go away in your suggestion, but be replaced by the same duplicated conversion inside of each operator)
3. Gets a result of unknown type and converts it into a literal value (likewise, this conversion could also be pushed down, but would be duplicated)

So the trade off that I see is: Is it better to take in make the functions `(literal[]) -> literal` (where literal is the literalvalue type currently in use that can hold a value of any type), and do the conversion inside each or `(object[]) -> object` (where object is the specific expected argument type/response type) and do all conversions in one spot outside. 

Ultimately, I would like to not expose the argument, but went with this approach for two reasons:
1. Just less repetitive. Even if sharing the coercion code itself, each projection gets more complex and those are the parts that I see changing/growing in the future.
2. In a future world (movie voice) where we are allowing defining attribute derivations on the fly, being able to lookup the metadata of the projection allows us to validate a definition and potentially build a UI around it. Admittedly, we'd need to build out more here potentially like number of args and return type which weren't needed right now.

Reply:
I believe I addressed both of my concerns in the current iteration, and moved the conversions down lower.